### PR TITLE
Custom items model proposal

### PR DESCRIPTION
**Pull request by 20kdc**
Making the "custom-items-model.js" thing more officially

### CODE REVIEW

**Comment by 20kdc** [comment]
NOTE: I might add a further detail in this proposal to make it work better with the patch-steps proposal.
Essentially, the idea behind this would be to allow custom items to be registered from JSON patches.
This solves feature holes in both specifications regarding the registration and usage of a custom item.

**Comment by 20kdc** [comment]
Actually, ignore that previous note.
The results wouldn't be that beneficial - for one, the item database isn't really suited to JSON patching in any form, and the stuff needed to fix that'd be out of scope for this particular specification.

File: proposals/1/custom-items-model.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+# Custom Items Model
+
+## Description Of Issue
+
+Eventually, story modifications will reach a point where items will need to be introduced.
+
+This is going to cause save compatibility issues.
```
Comment (2767mr) [inline proposals/1/custom-items-model.md:7]: Please describe the compatibility issues

File: proposals/1/custom-items-model.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+### Disadvantages
+
+- Any runtime generation of items that can occur after a savefile has been loaded may still lead to issues if double-allocation is not checked for.
+
+- It's possible future game versions will add other storage involving item IDs, or I may have missed places items could be lurking.
+
+- Should items be added by a modloader that does not conform to this specification, most of the positive effects of this specification are null and void if those items are introduced into the player's inventory. (This is probably inevitable.)
+
+- The vanilla game, or modded versions that don't conform to this specification, are likely to crash upon opening the Inventory menu of a save with custom items.
```
Comment (2767mr) [inline proposals/1/custom-items-model.md:33]: This is a **HUGE** negative. Maybe we should consider workarounds for this case.

File: proposals/1/custom-items-model.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+As such, modloaders that wish to inject the API must check for `sc.inventory.registerItemAsCustom`'s presence.
+
+If it exists, do NOT inject the API.
+
+The actual API is as follows:
+
+```
+sc.inventory.customItemToId : Object[String (custom item ID-names)] = Number (Item's ID in sc.inventory.items)
+sc.inventory.idToCustomItem : Object[Number (Item's ID in sc.inventory.items)] = String (custom item ID-names)
+sc.inventory.registerItemAsCustom : function (name, id) : Puts the custom item into the two previous objects.
```
Comment (2767mr) [inline proposals/1/custom-items-model.md:48]: Please describe the API with the TS declaration syntax

File: proposals/1/custom-items-model.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+```
+
+This must be implemented via `sc.Inventory.inject`.
+
+At the same time as the API is injected, a handler to transform save data must be injected.
+
+I recommend `sc.PlayerModel.getSaveData` and `sc.PlayerModel.preLoad` as the functions of choice to wrap.
```
Comment (2767mr) [inline proposals/1/custom-items-model.md:55]: The author of the spec should not recommend anything in a spec. Either rephrase it or make it a requiremet.

File: proposals/1/custom-items-model.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+`getSaveData`'s wrapping is simple - after receiving the resulting object from the parent function, add the key `"customItems"`, as a copy of `sc.inventory.customItemToId`.
```
Comment (2767mr) [inline proposals/1/custom-items-model.md:57]: "'s wrapping is simple" is subjective and should therefor not go into a spec

File: proposals/1/custom-items-model.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+- Remap all item references according to the mapping
+
+- Call the parent function with the remapped items
+
+## Reference Implementation
+
+A reference implementation as an Impact Module is available at:
+
+https://github.com/20kdc/CLS-20kdc-Code/tree/master/modules/game/feature/custom-items
```
Comment (2767mr) [inline proposals/1/custom-items-model.md:73]: Maybe we should move the ref implementation into this repo?

File: proposals/1/custom-items-model.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+Note that this doesn't contain the already-injected check.
```
Comment (2767mr) [inline proposals/1/custom-items-model.md:75]: Maybe rephrase this to be a bit more formal

**Review by 2767mr — CHANGES_REQUESTED** [review]
I don't know if it's a good idea to accept a proposal that would make saves incompatible with vanille. If a workaround is for that disadvantage, I would be fine with this content of this proposal.

**Comment by 20kdc** [comment]
This'll take a while to rewrite to be fully foolproof against breaking Vanilla, and Botanics might stir up some trouble anyway (Turns out drop counts won't be properly handled by the current setup - probably non-fatal but I still need to start checking everything to see where *else* all this data is flowing), so consider this on hold and/or cancelled.
